### PR TITLE
use reduced_precision_reduction flags in Triton matmul

### LINKED ISSUE #108621: [inductor] Triton matmul templates should use reduced_precision_reduction flags
In eager mode [we have flags](https://pytorch.org/docs/stable/notes/cuda.html#reduced-precision-reduction-in-fp16-gemms) for using fp16 accumulators in matmuls.

Currently, our Triton matmul templates ignore this flag and always accumulate in float32.  This will make them slower, so we may be leaving some perf on the table in max-autotune mode.

The implementation of this just involves updating the `acc_type` function to read the correct flags:
https://github.com/pytorch/pytorch/blob/c8e72a4a5c6398ad38d41ceee9775f8d4544225c/torch/_inductor/kernel/mm_common.py#L134-L137

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### PR DESCRIPTION
**Pull request by michiboo**
Fixes #108621

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/_inductor/kernel/mm_common.py [hunk 1]
```diff
@@ -133,6 +133,8 @@ def mm_grid(m, n, meta):
 
 def acc_type(dtype):
     if dtype in (torch.float16, torch.bfloat16):
+        if torch.backends.cuda.matmul.allow_fp16_reduced_precision_reduction:
+            return "tl.float16"
```
Comment (jansel) [inline torch/_inductor/kernel/mm_common.py:137]: I think we need to seperate out the float16 and bfloat16 cases because there is also:
```
torch.backends.cuda.matmul.allow_bf16_reduced_precision_reduction
```
in addition to this one.

**Review by jansel — CHANGES_REQUESTED** [review]
Can you add a test in https://github.com/pytorch/pytorch/blob/main/test/inductor/test_select_algorithm.py

@aakhundov also made a good comment about setting `out_dtype=` here https://github.com/pytorch/pytorch/issues/108621#issuecomment-1708064228

**Comment by michiboo** [comment]
> Can you add a test in https://github.com/pytorch/pytorch/blob/main/test/inductor/test_select_algorithm.py
> 
> @aakhundov also made a good comment about setting `out_dtype=` here [#108621 (comment)](https://github.com/pytorch/pytorch/issues/108621#issuecomment-1708064228)

Hello can you give some advise on how to test it?
Should I just turn on and off the flag and assert the dtype?

**Comment by aakhundov** [comment]
> Hello can you give some advise on how to test it? Should I just turn on and off the flag and assert the dtype?

I guess, one way would be to test the output of the PT Eager and `torch.compile`-d models for numerical equivalence under different reduced precision flags set.

File: torch/_inductor/kernel/mm_common.py (hunk 1 above)
Comment (jon-chuang) [reply torch/_inductor/kernel/mm_common.py:137]:
```python
cast_fp16 = dtype == torch.float16 and not torch.backends.cuda.matmul.allow_fp16_reduced_precision_reduction
cast_bf16 = dtype == torch.float16 and not torch.backends.cuda.matmul.allow_bf16_reduced_precision_reduction
if (cast_fp16 or cast_bf16):
  return "tl.float32"

**Comment by jon-chuang** [comment]
I got the following error without substitute `out_dtype` on Triton main. 
```
triton.compiler.errors.CompilationError: at 62:33:    # `accumulator` will be converted back to fp16 after the loop.
    accumulator = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_N), dtype=tl.float16)
    for k in range(0, tl.cdiv(K, BLOCK_SIZE_K)):
        # Load the next block of A and B, generate a mask by checking the K dimension.
        # If it is out of bounds, set it to 0.
        a = tl.load(a_ptrs, mask=offs_k[None, :] < K - k * BLOCK_SIZE_K, other=0.0)
        b = tl.load(b_ptrs, mask=offs_k[:, None] < K - k * BLOCK_SIZE_K, other=0.0)
        # We accumulate along the K dimension.
        accumulator += tl.dot(a, b, allow_tf32=True) #out_dtype=tl.float16)
        # Advance the ptrs to the next K block.
        a_ptrs += BLOCK_SIZE_K * stride_ak
        b_ptrs += BLOCK_SIZE_K * stride_bk
                                 ^
AssertionError('Loop-carried variable accumulator has initial type <[128, 256], fp16> but is re-assigned to <[128, 256], fp32> in loop! Please make sure that the type stays consistent.')
```

So not sure why the tests are passing

**Comment by michiboo** [comment]
> I got the following error without substitute `out_dtype` on Triton main.
> 
> ```
> triton.compiler.errors.CompilationError: at 62:33:    # `accumulator` will be converted back to fp16 after the loop.
>     accumulator = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_N), dtype=tl.float16)
>     for k in range(0, tl.cdiv(K, BLOCK_SIZE_K)):
>         # Load the next block of A and B, generate a mask by checking the K dimension.
>         # If it is out of bounds, set it to 0.
>         a = tl.load(a_ptrs, mask=offs_k[None, :] < K - k * BLOCK_SIZE_K, other=0.0)
>         b = tl.load(b_ptrs, mask=offs_k[:, None] < K - k * BLOCK_SIZE_K, other=0.0)
>         # We accumulate along the K dimension.
>         accumulator += tl.dot(a, b, allow_tf32=True) #out_dtype=tl.float16)
>         # Advance the ptrs to the next K block.
>         a_ptrs += BLOCK_SIZE_K * stride_ak
>         b_ptrs += BLOCK_SIZE_K * stride_bk
>                                  ^
> AssertionError('Loop-carried variable accumulator has initial type <[128, 256], fp16> but is re-assigned to <[128, 256], fp32> in loop! Please make sure that the type stays consistent.')
> ```
> 
> So not sure why the tests are passing

I think because due to this in test_select_algorithm.py, allow_tf32=false

    @functools.wraps(fn)
    def wrapped(*args, **kwargs):
        counters.clear()
        torch.manual_seed(12345)
        assert (
            not torch.backends.cuda.matmul.allow_tf32
        ), "correctness testing is allergic to tf32"
        return fn(*args, **kwargs)

**Comment by jon-chuang** [comment]
Btw, bf16 is not currently supported as an accumulation type in Triton. https://github.com/openai/triton/issues/1080

**Comment by jon-chuang** [comment]
I can add support shortly. But it will have to wait for later release of Triton.

**Comment by jon-chuang** [comment]
Btw, the test does _NOT_ test that the fp16/bf16 acc is used.

**Comment by jon-chuang** [comment]
We can add a test that the PTX emitted contains an mma_sync_fp16_fp16_fp16 op. But seems a bit overkill.

**Comment by jon-chuang** [comment]
> So not sure why the tests are passing

Seems like tests are skipped when the following is true:
https://github.com/pytorch/pytorch/blob/89eb7a75a251c41c4bee86e9ede1001b0d3998af/torch/_inductor/utils.py#L650

So not sure if CI is even running these tests

File: test/inductor/test_select_algorithm.py [hunk 2]
```diff
@@ -217,6 +217,36 @@ def foo(a):
         foo(torch.randn(32, 32, device="cuda"))
         self.check_counter(counters["inductor"]["select_algorithm_autotune"], 1)
 
+    @patch.object(select_algorithm, "VERIFY", dict(atol=5e-2, rtol=5e-2))
+    @patches
+    def test_mm_bf16_flags(self):
+        torch.backends.cuda.matmul.flags.allow_bf16_reduced_precision_reduction=True
```
Comment (jansel) [inline test/inductor/test_select_algorithm.py:223]: Move this to a `patch.object` at decorator so it gets cleaned up at the end.

File: test/inductor/test_select_algorithm.py [hunk 3]
```diff
@@ -217,6 +217,36 @@ def foo(a):
...
+            torch.randn(2, 8, 32, device="cuda", dtype=torch.bfloat16),
+            torch.randn(2, 32, 8, device="cuda", dtpye=torch.bfloat16),
+        )
+        
+        self.check_counter(counters["inductor"]["select_algorithm_autotune"], 1)
+
+    @patch.object(select_algorithm, "VERIFY", dict(atol=5e-2, rtol=5e-2))
+    @patches
+    def test_mm_f16_flags(self):
+        torch.backends.cuda.matmul.flags.allow_f16_reduced_precision_reduction=True
```
Comment (jansel) [inline test/inductor/test_select_algorithm.py:238]: Move this to a `patch.object` at decorator so it gets cleaned up at the end.

**Review by jansel — CHANGES_REQUESTED** [review]
If things aren't supported yet in Triton, let's just land what is supported and leave a comment in the code explaining it.

File: test/inductor/test_select_algorithm.py (hunk 2 above)
Comment (jon-chuang) [reply test/inductor/test_select_algorithm.py:223]: continuing here https://github.com/pytorch/pytorch/pull/108982/files#r1320786697 

This branch is a little off the mark, hope it's ok to take over @michiboo 

I noticed that you weren't even running tests locally.

**Comment by jon-chuang** [comment]
>  think because due to this in test_select_algorithm.py, allow_tf32=false

This is false, the tests were failing regardless

```
triton.compiler.errors.CompilationError: at 62:33:    # `accumulator` will be converted back to fp16 after the loop.
    accumulator = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_N), dtype=tl.float16)
    for k in range(0, tl.cdiv(K, BLOCK_SIZE_K)):
        # Load the next block of A and B, generate a mask by checking the K dimension.
        # If it is out of bounds, set it to 0.
        a = tl.load(a_ptrs, mask=offs_k[None, :] < K - k * BLOCK_SIZE_K, other=0.0)
        b = tl.load(b_ptrs, mask=offs_k[:, None] < K - k * BLOCK_SIZE_K, other=0.0)
        # We accumulate along the K dimension.
        accumulator += tl.dot(a, b, allow_tf32=False) #, out_dtype=tl.float16)
        # Advance the ptrs to the next K block.
        a_ptrs += BLOCK_SIZE_K * stride_ak
        b_ptrs += BLOCK_SIZE_K * stride_bk
                                 ^
AssertionError('Loop-carried variable accumulator has initial type <[128, 256], fp16> but is re-assigned to <[128, 256], fp32> in loop! Please make sure that the type stays consistent.')

```

File: test/inductor/test_select_algorithm.py (hunk 2 above)
Comment (michiboo) [reply test/inductor/test_select_algorithm.py:223]: @jon-chuang can you at least leave my commit on, cause I spend some time on this and as a zero time contributor, I would say it is not a good way to welcome a new member to the community.

Comment (jon-chuang) [reply test/inductor/test_select_algorithm.py:223]: if you really want you can clone the changes in my branch. 

It's just that your tests didn't run, was full of bugs, and the code change itself was one line, whereas it needed to touch a lot more files.

Furthermore, the tests weren't running on CI, so it was at risk of getting merged mistakenly as being correct.

I fixed enough issues that I think it warrants using my branch instead. I recommend trying to at least verify that your changes pass locally for future attempts.

File: test/inductor/test_select_algorithm.py (hunk 3 above)
Comment (jon-chuang) [inline test/inductor/test_select_algorithm.py:238]:
```diff
-torch.backends.cuda.matmul.flags.allow_f16_reduced_precision_reduction=True
+torch.backends.cuda.matmul.allow_fp16_reduced_precision_reduction=True
```

File: test/inductor/test_select_algorithm.py (hunk 2 above)
Comment (michiboo) [reply test/inductor/test_select_algorithm.py:223]: @jon-chuang First of all, this issue is labelled as "good first issue" that is why I choose to address it as first time contributor.
As such it is normal to have bugs for first time contributor, I think a little bit of patience will make the world a better place. If issue is more urgent then why labelled it as good first issue? 
I can see you are passionate about this issue, but next time can you please give some time for first time contributor on "beginner issue" and instead of creating separate branch on the issue, you can guide them on their mistakes and bugs.
I think that's more align to the open source values.

Comment (jon-chuang) [reply test/inductor/test_select_algorithm.py:223]: Well, apologies for my rashness. 

Feel free to merge changes from my branch into your branch, I was just rather alarmed by the number of errors and wanted to make sure it was all fixable.

Thankfully, the tests do indeed fail: https://github.com/pytorch/pytorch/actions/runs/6136917114/job/16654336560?pr=108637 so at least it is caught by CI.

File: test/test_ops.py [hunk 4]
```diff
@@ -648,7 +651,7 @@ def test_noncontiguous_samples(self, device, dtype, op):
     # Cases test here:
     #   - out= with the correct dtype and device, but the wrong shape
     @ops(_ops_and_refs, dtypes=OpDTypes.none)
-    @skipIfTorchInductor("Inductor does not support complex dtype yet")
+    @patch.object(torch, "manual_seed", deterministic_torch_manual_seed)
```
Comment (jon-chuang) [inline test/test_ops.py:654]: The test has now been marked as skipped in master. So we can remove the patch using `deterministic_torch_manual_seed`

File: benchmarks/dynamo/common.py [hunk 5]
```diff
@@ -375,17 +375,6 @@ def nothing(f):
 @functools.lru_cache(None)
 def patch_torch_manual_seed():
     """Make torch manual seed deterministic. Helps with accuracy testing."""
-
-    def deterministic_torch_manual_seed(*args, **kwargs):
```
Comment (jon-chuang) [inline benchmarks/dynamo/common.py:379]: remove these changes too. It will be fixed in a separate PR

File: torch/_inductor/utils.py [hunk 6]
```diff
@@ -732,10 +741,10 @@ def __len__(self):
 
 
 @functools.lru_cache(None)
-def is_big_gpu(index):
```
Comment (jon-chuang) [inline torch/_inductor/utils.py:735]: I'll try to move these changes into a separate PR after some investigation.

I have a feeling we can get rid of this check entirely.

File: torch/_inductor/utils.py [hunk 7]
```diff
@@ -764,9 +773,17 @@ def _use_autotune_backend(backend: str) -> bool:
 def use_triton_template(layout, *, enable_int32=False):
     layout_dtypes = [torch.float16, torch.bfloat16, torch.float32]
     if enable_int32:
-        layout_dtypes = [torch.float16, torch.bfloat16, torch.float32, torch.int32]
-    return _use_template_for_cuda(layout, layout_dtypes) and _use_autotune_backend(
-        "TRITON"
+        layout_dtypes.append(torch.int32)
+    return (
```
Comment (jon-chuang) [inline torch/_inductor/utils.py:777]: This has been moved into `_use_template_for_cuda`. Please revert these lines

File: benchmarks/dynamo/common.py (hunk 5 above)
Comment (michiboo) [reply benchmarks/dynamo/common.py:379]: regardless if it fixed in a separate PR, is it better to move the function out so it can be reuse from somewhere it needed?

Comment (jon-chuang) [reply benchmarks/dynamo/common.py:379]: Yes, good idea. However, it's better to have 1 PR do one thing. I'll move into common util in https://github.com/pytorch/pytorch/pull/108998

File: torch/_inductor/utils.py (hunk 6 above)
Comment (jon-chuang) [reply torch/_inductor/utils.py:735]: There is a PR and ongoing investigation now: https://github.com/pytorch/pytorch/pull/109492

**Comment by jon-chuang** [comment]
@michiboo are you still working on this? Else I will take over.

**Comment by jansel** [comment]
@jon-chuang since there is no update in the past month, please take over.
